python/tank/authentication/user.py

```diff
@@ -118,6 +126,93 @@ def impl(self):
         return self._impl
 
 
+class ShotgunSamlUser(ShotgunUser):
+    """
+    This specialized shotgun user is needed when SSO is used, as it provides
+    mechanisms
+    """
+
+    def __init__(self, impl):
+        """
+        :param impl: Internal user implementation class this class proxies.
+        """
+        super(ShotgunSamlUser, self).__init__(impl)
+        self._timer = None
+
+    def get_claims_expiration(self):
+        """
+        Obtain claims expiration for the user.
+
+        :returns: The expiration in seconds since January 1st 1970 UTC.
+        """
+        # The import must be located here. If it is at global scope, there
+        # is an import error.
+        from .shotgun_shared import saml2_sso
+
+        return saml2_sso.get_saml_claims_expiration(self._impl.get_cookies())
+
+    def _automatic_claims_renewal(self, preemtive_renewal_threshold=0.9):
+        """
+        Handles automatic renewal of the SAML2 claims for the user.
+
+        :params user: an already logged in user.
+        :params preemtive_renewal_threshold: How far into the claims duration we will attempt renewal.
+                                             Defaults to 90%, usually 3 minutes 45 seconds (90% of 5 mins).
+        """
+        from tank.authentication import interactive_authentication
+        from sgtk.authentication import AuthenticationCancelled
+
+        logger.debug("Automatic claims renewal")
+        try:
+            previous_expiration = self.get_claims_expiration()
+            # A call to renew_session when SSO is used will not prompt the user for
+            # their credentials if it is not necessary.
+            interactive_authentication.renew_session(self._impl)
+            new_expiration = self.get_claims_expiration()
+
+            if new_expiration > previous_expiration:
+
+                logger.debug("Automatic claims renewal succeeded.")
+                delta = (new_expiration - time.time()) * preemtive_renewal_threshold
+                # If we are debugging, we will use a shorter expiration time.
+                if "SHOTGUN_SSO_DEVELOPER_ENABLED" in os.environ:
```

Comment:
We had that discussion before in another pull request review : there is no need to tweak the value, the desired effect is to cause the renewal to happen more often. (e.g. in a matter of seconds instead of minutes). Setting it to 5 seconds achieves the desired result.

But I will make the change nonetheless.